applylib/applyset/applyset.go

```diff
@@ -153,11 +242,151 @@ func (a *ApplySet) ApplyOnce(ctx context.Context) (*ApplyResults, error) {
+	if to == nil {
+		to = make(map[string]string)
+	}
+	for k, v := range from {
+		to[k] = v
+	}
+	return to
+}
+
+// updateLabel adds the "applyset.kubernetes.io/part-of: Parent-ID" label to the manifest.
+func (a *ApplySet) updateManifestLabel(obj ApplyableObject, applysetLabels map[string]string) error {
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("unable to convert `ApplyableObject` to `unstructured.Unstructured` %v/%v %v",
+			obj.GetName(), obj.GetNamespace(), obj.GroupVersionKind().String())
+	}
+	labels := u.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	newLabels := mergeMap(applysetLabels, labels)
+	u.SetLabels(newLabels)
+	return nil
+}
+
+// updateParentLabelsAndAnnotations updates the parent labels and annotations.
+// This method leverages the kubectlapply to build the parent labels and annotations, but avoid using the
+// `resource.NewHelper` and cmdutil to patch the parent.
+func (a *ApplySet) updateParentLabelsAndAnnotations(ctx context.Context, kapplyset *kubectlapply.ApplySet, mode kubectlapply.ApplySetUpdateMode) error {
+	parent, err := meta.Accessor(a.parent.GetSubject())
+	if err != nil {
+		return err
+	}
+
+	original, err := meta.Accessor(a.parent.GetSubject().DeepCopyObject())
+	if err != nil {
+		return err
+	}
+	partialParent := kapplyset.BuildParentPatch(mode)
+	newAnnotations := mergeMap(partialParent.Annotations, parent.GetAnnotations())
+	parent.SetAnnotations(newAnnotations)
+	newLabels := mergeMap(partialParent.Labels, parent.GetLabels())
+	parent.SetLabels(newLabels)
+
+	return a.updateParent(ctx, original, parent)
+}
+
+func (a *ApplySet) deleteObjects(ctx context.Context, pruneObjects []kubectlapply.PruneObject, results *ApplyResults) error {
+	for i := range pruneObjects {
+		pruneObject := &pruneObjects[i]
+
+		name := pruneObject.Name
+		namespace := pruneObject.Namespace
+		mapping := pruneObject.Mapping
+		gvk := pruneObject.Object.GetObjectKind().GroupVersionKind()
+		nn := types.NamespacedName{Namespace: namespace, Name: name}
+
+		if err := a.client.Resource(mapping.Resource).Namespace(namespace).Delete(ctx, name, a.deleteOptions); err != nil {
+			klog.Error("unable to delete resource ")
+			results.pruneError(gvk, nn, fmt.Errorf("error from delete: %w", err))
+		} else {
+			klog.Infof("pruned resource %v", pruneObject.String())
+			results.pruneSuccess(gvk, nn)
+		}
+	}
+	return nil
+}
+
+func (a *ApplySet) WithParent(ctx context.Context, kapplyset *kubectlapply.ApplySet) error {
+	parent := a.parent.GetSubject()
+	object, err := meta.Accessor(parent)
+	if err != nil {
+		return err
+	}
+	original, err := meta.Accessor(parent.DeepCopyObject())
+	if err != nil {
+		return err
+	}
+	//kubectlapply requires the tooling and id to exist.
+	{
+		annotations := object.GetAnnotations()
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[kubectlapply.ApplySetToolingAnnotation] = a.tooling
```

Comment:
Yes, I find this much easier to read than mergeMap

Reply:
😄  Sure.